contracts/src/v0.8/functions/dev/1_0_0/RouterBase.sol

```diff
@@ -0,0 +1,388 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import {IRouterBase} from "./interfaces/IRouterBase.sol";
+import {ConfirmedOwnerWithProposal} from "../../../ConfirmedOwnerWithProposal.sol";
+import {ITypeAndVersion} from "../../../shared/interfaces/ITypeAndVersion.sol";
+import {Pausable} from "../../../shared/vendor/openzeppelin-solidity/v.4.8.0/contracts/security/Pausable.sol";
+import {IConfigurable} from "./interfaces/IConfigurable.sol";
+
+abstract contract RouterBase is IRouterBase, Pausable, ITypeAndVersion, ConfirmedOwnerWithProposal {
+  // ================================================================
+  // |                         Version state                        |
+  // ================================================================
+  uint16 internal constant s_majorVersion = 1;
+  uint16 internal s_minorVersion = 0;
+  uint16 internal s_patchVersion = 0;
+
+  // ================================================================
+  // |                          Route state                         |
+  // ================================================================
+  mapping(bytes32 => address) internal s_route; /* jobId => contract address */
+  error RouteNotFound(bytes32 jobId);
+
+  // ================================================================
+  // |                         Proposal state                       |
+  // ================================================================
+  uint8 internal constant MAX_PROPOSAL_SET_LENGTH = 8;
+
+  struct ProposalSet {
+    bytes32[] jobIds;
+    address[] from;
+    address[] to;
+    uint proposedAtBlock;
+  }
+  ProposalSet internal s_proposalSet;
+
+  event Proposed(
+    bytes32[] proposalSetJobIds,
+    address[] proposalSetFromAddresses,
+    address[] proposalSetToAddresses,
+    uint block
+  );
+  event Upgraded(
+    bytes32[] proposalSetJobIds,
```

Comment:
For observability it is better to linearize these events. E.g. instead of one event with arrays, emit N events, where N is the array size. This way we could model DB structure to have don/from/to addresses in one row. This simplifies SQL queries that need the distinct set of all addresses in history. I suspect this query will be executed in every benthos app.
I believe in most cases the size of array would be 1. Therefore, let's optimize for this.